tree/ntuple/v7/src/RField.cxx

```diff
@@ -1083,6 +1083,211 @@ void ROOT::Experimental::RVectorField::AcceptVisitor(Detail::RFieldVisitor &visi
+   // this might not be the case if T in RVec<T> is over-aligned.
+   auto *buffer = new char[nItems * fItemSize];
+   for (auto i = 0u; i < nItems; ++i) {
+      auto itemValue = fSubFields[0]->GenerateValue(buffer + (i * fItemSize));
+      fSubFields[0]->Read(collectionStart + i, &itemValue);
+   }
+
+   void **rvecAddr = reinterpret_cast<void**>(value->GetRawPtr());
+   // void* fBegin is the first data member
+   *rvecAddr = buffer;
+   // int32_t fSize is the second data member (after 1 void*)
+   int32_t *size = reinterpret_cast<int32_t *>(rvecAddr + 1);
+   *size = nItems;
+   int32_t *capacity = size + 1;
+   *capacity = nItems;
+}
+
+void ROOT::Experimental::RRVecField::GenerateColumnsImpl()
+{
+   RColumnModel modelIndex(EColumnType::kIndex, true /* isSorted*/);
+   fColumns.emplace_back(std::unique_ptr<Detail::RColumn>(
+      Detail::RColumn::Create<ClusterSize_t, EColumnType::kIndex>(modelIndex, 0)));
+}
+
+void ROOT::Experimental::RRVecField::GenerateColumnsImpl(const RNTupleDescriptor &desc)
+{
+   EnsureColumnType({EColumnType::kIndex}, 0, desc);
+   GenerateColumnsImpl();
+}
+
+ROOT::Experimental::Detail::RFieldValue ROOT::Experimental::RRVecField::GenerateValue(void* where)
+{
+   // initialize data members fBegin, fSize, fCapacity
+   // currently the inline buffer is left uninitialized
+   void **beginPtr = new (where) (void *)(nullptr);
+   int32_t* sizePtr = new (reinterpret_cast<void*>(beginPtr + 1)) int32_t(0);
+   new (sizePtr + 1) int32_t(0);
+
+   return Detail::RFieldValue(/*captureTag*/true, this, where);
+}
+
+void ROOT::Experimental::RRVecField::DestroyValue(const Detail::RFieldValue& value, bool dtorOnly)
+{
+   void **rvecAddr = reinterpret_cast<void**>(value.GetRawPtr());
+   // void* fBegin is the first data member
+   void *begin = *rvecAddr;
+   // int32_t fSize is the second data member (after 1 void*, no padding expected)
+   const int32_t* sizePtr = reinterpret_cast<int32_t *>(rvecAddr + 1);
+   const int32_t size = *sizePtr;
+   R__ASSERT(size >= 0);
+   // int32_t fCapacity is the third data member (after fSize, no padding expected)
+   const int32_t capacity = *(sizePtr + 1);
+
+   for (int32_t i = 0; i < size; ++i) {
+      auto elementValue = fSubFields[0]->CaptureValue(reinterpret_cast<void *>(std::intptr_t(begin) + i * fItemSize));
+      fSubFields[0]->DestroyValue(elementValue, true /* dtorOnly */);
+   }
+
+   // figure out if we are in the small state, i.e. begin == &inlineBuffer
+   // there might be padding between fCapacity and the inline buffer, so we compute it here
+   const auto dataMemberSz = sizeof(void *) + 2 * sizeof(int32_t);
+   const auto alignOfT = fSubFields[0]->GetAlignment();
+   auto paddingMiddle = dataMemberSz % alignOfT;
+   if (paddingMiddle != 0)
+      paddingMiddle = alignOfT - paddingMiddle;
+   const bool isSmall = begin == (rvecAddr + dataMemberSz + paddingMiddle);
+
+   const bool owns = capacity != -1;
+   if (!dtorOnly && !isSmall && owns)
+      free(begin);
```

Comment:
Are we sure that the memory pointed to by `begin` was allocated by `malloc` (not by `new[]`)? E.g. line 1130 doesn't.

Reply:
ah, you are right. Is `RField::DestroyValue` with `!dtorOnly` always called on memory that `RField` allocated? If yes, I can just make line 1130 a malloc. If not, I'm not sure how to proceed

Reply:
will double-check but should be ok